pkg/pattern/metric/push.go

```diff
@@ -0,0 +1,256 @@
+type Push struct {
+	lokiURL     string
+	tenantID    string
+	httpClient  *http.Client
+	userAgent   string
+	contentType string
+	logger      log.Logger
+
+	// channel for incoming logs
+	entries chan entry
+
+	// shutdown channels
+	quit chan struct{}
+	done chan struct{}
+
+	// auth
+	username, password string
+
+	// Will add these label to the logs pushed to loki
+	labelName, labelValue, streamName, streamValue string
+
+	// push retry and backoff
+	backoff *backoff.Config
+}
+
+type entry struct {
+	ts     time.Time
+	entry  string
+	labels labels.Labels
+}
+
+// NewPush creates an instance of `Push` which writes logs directly to given `lokiAddr`
+func NewPush(
+	lokiAddr, tenantID string,
+	timeout time.Duration,
+	cfg config.HTTPClientConfig,
+	username, password string,
+	useTLS bool,
+	backoffCfg *backoff.Config,
+	logger log.Logger,
+) (*Push, error) {
+	client, err := config.NewClientFromConfig(cfg, "pattern-ingester-push", config.WithHTTP2Disabled())
+	if err != nil {
+		return nil, err
+	}
+
+	client.Timeout = timeout
+	scheme := "http"
+
+	// setup tls transport
+	if useTLS {
+		scheme = "https"
+	}
+
+	u := url.URL{
+		Scheme: scheme,
+		Host:   lokiAddr,
+		Path:   pushEndpoint,
+	}
+
+	p := &Push{
+		lokiURL:     u.String(),
+		tenantID:    tenantID,
+		httpClient:  client,
+		userAgent:   defaultUserAgent,
+		contentType: defaultContentType,
+		username:    username,
+		password:    password,
+		logger:      logger,
+		entries:     make(chan entry, 50), // Use a buffered channel so we can retry failed pushes without blocking WriteEntry
+		quit:        make(chan struct{}),
+		done:        make(chan struct{}),
+		backoff:     backoffCfg,
+	}
+	go p.run()
+	return p, nil
+}
+
+// WriteEntry implements EntryWriter
+func (p *Push) WriteEntry(ts time.Time, e string, lbls labels.Labels) {
+	p.entries <- entry{ts, e, lbls}
+}
+
+// Stop will cancel any ongoing requests and stop the goroutine listening for requests
+func (p *Push) Stop() {
+	if p.quit != nil {
+		close(p.quit)
+		<-p.done
+		p.quit = nil
+	}
+}
+
+// buildPayload creates the snappy compressed protobuf to send to Loki
+func (p *Push) buildPayload(e entry) ([]byte, error) {
+	req := &logproto.PushRequest{
+		Streams: []logproto.Stream{
+			{
+				Labels: e.labels.String(),
+				Entries: []logproto.Entry{
+					{
+						Timestamp: e.ts,
+						Line:      e.entry,
+					},
+				},
+				Hash: e.labels.Hash(),
+			},
+		},
+	}
+	payload, err := proto.Marshal(req)
+	if err != nil {
+		return []byte{}, fmt.Errorf("failed to marshal payload to json: %w", err)
+	}
+
+	payload = snappy.Encode(nil, payload)
+
+	return payload, nil
+}
+
+// run pulls lines out of the channel and sends them to Loki
+func (p *Push) run() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer func() {
+		close(p.done)
+	}()
+
+	for {
+		select {
+		case <-p.quit:
+			cancel()
+			return
+			// TODO(twhitney): could buffer these into fewer pushes?
```

Comment:
We can't push line by line !

Reply:
we also need to increment a single `int64` for each push, rather than storing all the raw samples. I can make both those changes here, or follow up in a subsequent PR?

Reply:
another option would be to use something like a buffered channel and call the API when the channel is full?